### ISSUE TITLE
Pursue Solr commit-within performance gains

### ISSUE THREAD

**Issue by atz** [sul-dlss/argo#96]
Use Solr `commit-within` parameters to try to speed indexing.  Possibly gets pushed down to dor-services.

**Comment by drh-stanford** [comment]
The `/dor/reindex/:id` route calls the reindex method which is using a `commit-within` parameter:

  https://github.com/sul-dlss/argo/blob/develop/app/controllers/dor_controller.rb#L50

There's another reindexing task here which also uses the `commit-within` parameter:

  https://github.com/sul-dlss/argo/blob/develop/lib/tasks/argo.rake#L373

Can you be more specific on where the `commit-within` parameter would be added?

Also, there's some history on prior attempts to improve indexing speed from the perspective of the Fedora update notification messages here: https://github.com/sul-dlss/workflow-multiplexing/issues/67

**Comment by atz** [comment]
The parameter is milliseconds, so the values of `1000` and `10` are likely too small to yield useful benefits.  In particular, 10ms is counter-productively small.

**Comment by atz** [comment]
@lmcglohon reports searchworks saw its benefits from a setting at `10000`.  That's probably where we should be.

**Comment by drh-stanford** [comment]
this issue should migrate to dor-indexing-app. any remaining indexing done by Argo should not have a commitWithin parameter, right?

**Comment by jmartin-sul** [comment]
well, `Dor::IndexingService` doesn't use `commitWithin` (because `Argo::Indexer` didn't, because we were skeptical of its benefit on single object indexing at the time that the argo indexing code was first centralized, esp with the low value it had).  but we could make that an optional thing so that something like bulk reindexing could specify a long `commitWithin` parameter value.

i did a quick grep for `commitWithin` (which appears to be the way it's spelled based on a quick look at the docs, i'm not particularly familiar with its usage myself:  https://wiki.apache.org/solr/UpdateXmlMessages#Passing_commit_and_commitWithin_parameters_as_part_of_the_URL).  it appears only once in argo, commented out in registration_helper.rb:
```ruby
  # @param [String] druid unqualified DRUID identifier
  # @return [Hash] doc from Solr or to_solr
  # @note To the extent we use Solr input filters or copyField(s), the Solr version will differ from the to_solr hash.
  # @note That difference shouldn't be important for the few known fields we use here.
  def find_or_create_in_solr_by_id(druid)
    doc = Dor::SearchService.query(%(id:"druid:#{druid}"), :rows => 1).docs.first
    return doc unless doc.nil?
    obj = Dor.load_instance "druid:#{druid}"
    solr_doc = obj.to_solr
    Dor::SearchService.solr.add(solr_doc)
    # Dor::SearchService.solr.add(solr_doc, :add_attributes => {:commitWithin => 1000})
    solr_doc
  end
```

it seems like that should get cleaned up to use the centralized dor-services indexing code anyway:
https://github.com/sul-dlss/dor-services/blob/develop/lib/dor/services/indexing_service.rb

**Comment by jmartin-sul** [comment]
which is to say, if this issue migrates anywhere, it should probably be to dor-services first, since it should optionally allow `commitWithin` for `reindex_object` and `reindex_pid`:
https://github.com/sul-dlss/dor-services/blob/develop/lib/dor/services/indexing_service.rb#L25

and then downstream consumers can take advantage of the option as desired.

**Comment by drh-stanford** [comment]
i vote to close the issue since we're not using `commitWithin` now. the commit window behavior is better left as configuration in the Solr core administration imho. See https://cwiki.apache.org/confluence/display/solr/UpdateHandlers+in+SolrConfig

**Comment by drh-stanford** [comment]
I created a new issue https://github.com/sul-dlss/argo/issues/515 for the code migration. and closing this issue re: commitWithin. please reopen we want to support the commitWithin parameter after all.
